alica_tracing/include/tracing/AlicaTrace.h

```diff
@@ -0,0 +1,197 @@
+    }
+
+    // Tags are the key value pair you can search in tracing ui
+    // If there is a pre-existing tag set for `key`, it is overwritten.
+    // Tag values can be numeric types, strings, or bools.
+    void setTag(const std::string& key, const std::string& value)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->SetTag(key, value);
+    }
+
+    // setLog is a timestamped way to record key:value logging data
+    // about a trace. Here's an example:
+    //
+    //    trace.Log({
+    //        {"progress", "checkpoint 4 cleared"},
+    //        {"wait.time", "waited 10 sec for agent 2"}});
+    // BE CAREFUL about what you want to log,
+    // These are supposed to be micro logs to be carried over the network.
+    void setLog(std::pair<std::string, std::string> logEntry)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->Log({logEntry});
+    }
+
+    // When operation being traced by this instance fail, call this api
+    // Error traces are highlighted in tracing ui.
+    void markError(const std::string& description)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->SetTag("error", true);
+        _rawTrace->SetTag("error.description", description);
+    }
+
+    // Get the context of this trace to propogate across process boundary
+    std::string context() const
+    {
+        if (!_rawTrace) {
+            return "";
+        }
+
+        const auto& jctx = dynamic_cast<const jaegertracing::SpanContext&>(_rawTrace->context());
+        std::stringstream ss;
+        jctx.print(ss);
+        return ss.str();
+    }
+
+private:
+    // This could be intentionally empty in the case when factory isn't initialized.
+    // An app can choose not to initialize the TraceFactory and
+    // as a result, traces should become no-op.
+    RawTracePtr _rawTrace;
+};
+
+using TracePtr = std::unique_ptr<Trace>;
+
+template <class TraceType>
+class TraceFactory : public alica::IAlicaTraceFactory
+{
+public:
+    TraceFactory() = default;
+
+    TraceFactory(const TraceFactory&) = delete;
+    TraceFactory(TraceFactory&&) = delete;
+    TraceFactory& operator=(const TraceFactory&) = delete;
+    TraceFactory&& operator=(TraceFactory&&) = delete;
+
+    ~TraceFactory()
+    {
+        ROS_INFO_STREAM_NAMED(__func__, "Terminating tracing for service " << _serviceName);
+        // allow termination to propogate,
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        opentracing::Tracer::Global()->Close();
+    };
+
+    // Initialize the trace factory once per process
+    // serviceName is the component running the instance, ex: lbc, gbc
+    // configFilePath is the place factory will look out for tracing configuration.
+    // returns false if initialization failed, true otherwise
+    // Will throw if the initialization failed
+    void init(const std::string& serviceName, const std::string& configFilePath)
```

Comment:
need to define a constructor that takes these parameters & does the initilization. The constructor should also take in the default tags as a parameter & set them after the initialization. Otherwise it won't be possible for the AlicaContext to create the factory instance internally.